File: javascript/2015/day05b/day05b.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const process = require('process');

function usage() {
    const progname = path.basename(process.argv[1]);
    console.log(`usage: ${progname} <input file>`);
    process.exit(1);
}

function prop1(str) {
    return /(..).*\1/.test(str);
}

function prop2(str) {
    return /(.).\1/.test(str);
}

function processFile(filename) {
    let count = 0;
    const data = fs.readFileSync(filename, 'utf8');
    const lines = data.trim().split('\n');
    for ( const line of lines ) {
        if ( prop1(line) && prop2(line) ) {
            count++;
        }
    }
    return count;
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    const filename = process.argv[2];
    const result = processFile(filename);
    console.log(`result = ${result}`);
}

main()

File: dart/2015/day05b/day05b.dart
#!/usr/bin/env dart

import 'dart:io';

void usage(String progName) {
    print('usage: $progName <input file>');
    exit(1);
}

bool prop1(String str) {
    for ( int i = 0; i < str.length - 3; i++ ) {
        String first = str.substring(i, i + 2);
        for ( int j = i + 2; j < str.length - 1; j++ ) {
            String second = str.substring(j, j + 2);
            if ( first == second ) {
                return true;
            }
        }
    }
    return false;
}

bool prop2(String str) {
    for ( int i = 2; i < str.length; i++ ) {
        if ( str[i] == str[i - 2] ) {
            return true;
        }
    }
    return false;
}

int process(String filename) {
    int count = 0;

    final lines = File(filename).readAsLinesSync();
    for ( final line in lines ) {
        if ( prop1(line) && prop2(line) ) {
            count++;
        }
    }
    return count;
}

void main(List<String> args) {
    if ( args.isEmpty ) {
        usage(Platform.script.pathSegments.last);
    }
    final filename = args[0];
    final result = process(filename);
    print('result = $result');
}
